timelock-agent/src/main/java/com/palantir/timelock/paxos/PaxosAgent.java

```diff
@@ -0,0 +1,124 @@
+import com.palantir.timelock.config.PaxosInstallConfiguration;
+import com.palantir.timelock.config.PaxosRuntimeConfiguration;
+import com.palantir.timelock.config.TimeLockDeprecatedConfiguration;
+import com.palantir.timelock.config.TimeLockInstallConfiguration;
+import com.palantir.timelock.config.TimeLockRuntimeConfiguration;
+
+import io.reactivex.Observable;
+
+public class PaxosAgent extends TimeLockAgent {
+    private final PaxosInstallConfiguration paxosInstall;
+
+    private final PaxosResource paxosResource;
+    private final PaxosLeadershipCreator leadershipAgent;
+    private final LockCreator lockCreator;
+    private final PaxosTimestampCreator timestampCreator;
+
+    public PaxosAgent(TimeLockInstallConfiguration install,
+            Observable<TimeLockRuntimeConfiguration> runtime,
+            Consumer<Object> registrar) {
+        this(install, runtime, ImmutableTimeLockDeprecatedConfiguration.builder().build(), registrar);
+    }
+
+    public PaxosAgent(TimeLockInstallConfiguration install,
+            Observable<TimeLockRuntimeConfiguration> runtime,
+            TimeLockDeprecatedConfiguration deprecated,
+            Consumer<Object> registrar) {
+        super(install, runtime, deprecated, registrar);
+        Preconditions.checkState(install.algorithm() instanceof PaxosInstallConfiguration,
+                "Cannot initialize a Paxos agent with a non-Paxos algorithm configuration.");
+        this.paxosInstall = (PaxosInstallConfiguration) install.algorithm();
+
+        this.paxosResource = PaxosResource.create(paxosInstall.dataDirectory().toString());
+        this.leadershipAgent = new PaxosLeadershipCreator(install, runtime, registrar);
+        this.lockCreator = new LockCreator(runtime, deprecated);
+        this.timestampCreator = new PaxosTimestampCreator(paxosResource,
+                PaxosRemotingUtils.getRemoteServerPaths(install),
+                install.cluster().cluster().security().map(SslSocketFactories::createSslSocketFactory),
+                runtime.map(x -> x.algorithm().orElse(ImmutablePaxosRuntimeConfiguration.builder().build())));
+    }
+
+    @Override
+    public void createAndRegisterResources() {
+        registerPaxosResource();
+        registerPaxosExceptionMappers();
+        leadershipAgent.registerLeaderElectionService();
+
+        // Finally, register the endpoints associated with the clients.
+        super.createAndRegisterResources();
+    }
+
+    // No runtime configuration at the moment.
+    private void registerPaxosResource() {
+        registrar.accept(paxosResource);
+    }
+
+    private void registerPaxosExceptionMappers() {
+        registrar.accept(new BlockingTimeoutExceptionMapper());
+        registrar.accept(new NotCurrentLeaderExceptionMapper());
+        registrar.accept(new TooManyRequestsExceptionMapper());
+    }
+
+    @Override
+    public TimeLockServices createInvalidatingTimeLockServices(String client) {
```

Comment:
Currently, the generic TimeLockAgent on startup calls this method for each client, and builds the map for the `TimeLockResource` (the jersey resource locator that sends calls off to each timestamp and lock service) based on this.

Eventually, I can still see this being used to handle the logic that says "create timestamp and lock services for a given client". Though I'm not too sure how this will be invoked when we do live reloading. 

We could start by registering the available clients known on startup, based off the runtime config... though not sure there's much of an advantage to doing that as opposed to starting with 0 and treating the first observed config as a change to add all the clients.